streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignorTest.java

```diff
@@ -0,0 +1,492 @@
+        secondAssignor.assign(0);
+
+        assertThat(clients.get(p2).activeTasks(), equalTo(Collections.singleton(task00)));
+        assertThat(clients.get(p4).activeTasks(), equalTo(Collections.singleton(task02)));
+        assertThat(clients.get(p5).activeTasks(), equalTo(Collections.singleton(task01)));
+
+
+    }
+
+    @Test
+    public void shouldAssignTasksToClientWithPreviousStandbyTasks() throws Exception {
+        final ClientState<TaskId> client1 = createClientWithPreviousActiveTasks(p1, 1);
+        client1.addPreviousStandbyTasks(Utils.mkSet(task02));
+        final ClientState<TaskId> client2 = createClientWithPreviousActiveTasks(p2, 1);
+        client2.addPreviousStandbyTasks(Utils.mkSet(task01));
+        final ClientState<TaskId> client3 = createClientWithPreviousActiveTasks(p3, 1);
+        client3.addPreviousStandbyTasks(Utils.mkSet(task00));
+
+        final StickyTaskAssignor taskAssignor = createTaskAssignor(task00, task01, task02);
+
+        taskAssignor.assign(0);
+
+        assertThat(clients.get(p1).activeTasks(), equalTo(Collections.singleton(task02)));
+        assertThat(clients.get(p2).activeTasks(), equalTo(Collections.singleton(task01)));
+        assertThat(clients.get(p3).activeTasks(), equalTo(Collections.singleton(task00)));
+    }
+
+    @Test
+    public void shouldAssignBasedOnCapacityWhenMultipleClientHaveStandbyTasks() throws Exception {
+        final ClientState<TaskId> c1 = createClientWithPreviousActiveTasks(p1, 1, task00);
+        c1.addPreviousStandbyTasks(Utils.mkSet(task01));
+        final ClientState<TaskId> c2 = createClientWithPreviousActiveTasks(p2, 2, task02);
+        c2.addPreviousStandbyTasks(Utils.mkSet(task01));
+
+        final StickyTaskAssignor taskAssignor = createTaskAssignor(task00, task01, task02);
+
+        taskAssignor.assign(0);
+
+        assertThat(clients.get(p1).activeTasks(), equalTo(Collections.singleton(task00)));
+        assertThat(clients.get(p2).activeTasks(), equalTo(Utils.mkSet(task02, task01)));
+    }
+
+    @Test
+    public void shouldAssignStandbyTasksToDifferentClientThanCorrespondingActiveTaskIsAssingedTo() throws Exception {
+        createClientWithPreviousActiveTasks(p1, 1, task00);
+        createClientWithPreviousActiveTasks(p2, 1, task01);
+        createClientWithPreviousActiveTasks(p3, 1, task02);
+        createClientWithPreviousActiveTasks(p4, 1, task03);
+
+        final StickyTaskAssignor taskAssignor = createTaskAssignor(task00, task01, task02, task03);
+        taskAssignor.assign(1);
+
+        assertThat(clients.get(p1).standbyTasks(), not(hasItems(task00)));
+        assertThat(clients.get(p2).standbyTasks(), not(hasItems(task01)));
+        assertThat(clients.get(p3).standbyTasks(), not(hasItems(task02)));
+        assertThat(clients.get(p4).standbyTasks(), not(hasItems(task03)));
+        assertThat(allStandbyTasks(), equalTo(Arrays.asList(task00, task01, task02, task03)));
```

Comment:
It is not always possible (with the current algo) to guarantee that each task will get 1 task. It depends somewhat on the previous active tasks and the order that the tasks are iterated in. I can change the test to make sure that at least 3 tasks have standby and that no task has more than 2